internal/definitions/sender_tokenpool.go

```diff
@@ -24,34 +24,91 @@ import (
 	"github.com/hyperledger/firefly-common/pkg/log"
 	"github.com/hyperledger/firefly/internal/coremsgs"
 	"github.com/hyperledger/firefly/pkg/core"
+	"github.com/hyperledger/firefly/pkg/database"
 )
 
-func (bm *definitionSender) DefineTokenPool(ctx context.Context, pool *core.TokenPoolAnnouncement, waitConfirm bool) error {
+func (ds *definitionSender) PublishTokenPool(ctx context.Context, poolNameOrID, networkName string, waitConfirm bool) (pool *core.TokenPool, err error) {
+	if !ds.multiparty {
+		return nil, i18n.NewError(ctx, coremsgs.MsgActionNotSupported)
+	}
 
-	if bm.multiparty {
-		// Map token connector name -> broadcast name
-		if broadcastName, exists := bm.tokenBroadcastNames[pool.Pool.Connector]; exists {
-			pool.Pool.Connector = broadcastName
-		} else {
-			log.L(ctx).Infof("Could not find broadcast name for token connector: %s", pool.Pool.Connector)
-			return i18n.NewError(ctx, coremsgs.MsgInvalidConnectorName, broadcastName, "token")
+	var sender *sendWrapper
+	err = ds.database.RunAsGroup(ctx, func(ctx context.Context) error {
+		if pool, err = ds.assets.GetTokenPoolByNameOrID(ctx, poolNameOrID); err != nil {
+			return err
+		}
+		if networkName != "" {
+			pool.NetworkName = networkName
 		}
 
-		if err := pool.Pool.Validate(ctx); err != nil {
-			return err
+		sender = ds.getTokenPoolSender(ctx, pool)
+		if sender.err != nil {
+			return sender.err
 		}
+		if !waitConfirm {
```

Comment:
I'm trying to understand/remember why we change the order of persistence based on whether we engaging the sync/async bridge or not. Bit worried these divergent codepaths might leave the door open to different behavior.

Reply:
I'm trying to page back in as well... I think because this method is only called from the API server, there are two possible outcomes:
1. I requested async, we have to persist the token pool here in order to add the message ID
2. I requested sync, the token pool is going to get persisted anyway before I respond (ie by the event loop), so we skip this upfront save

I don't _think_ there's any problem here, but I acknowledge it's probably an over-optimization in (2). Perhaps cleaner to always Prepare+Upsert upfront?

Reply:
I guess in the case that the message in (2) is lost or stuck, the token pool <-> message linkage might not be persisted to the database, and that would be problematic at least for debugging.

Reply:
After further pondering, I've decided we shouldn't Upsert with the message ID at all. We need to wait until the message is received/confirmed. In the (rare) case of a race, it should be the winning/first message that gets persisted onto the token pool object. If we persist the ID of an unconfirmed message, it becomes more complex to handle all the possible branches when a message is later confirmed.